File: java/src/main/java/behavioral/ChainOfResponsibility.java
package behavioral;

abstract class Handler {

    protected Handler next;

    public Handler setNext(Handler next) {
        this.next = next;
        return next;
    }

    public void handle(String request) {
        if (next != null) {
            next.handle(request);
        }
    }
}

// Concrete handlers
class AuthHandler extends Handler {

    @Override
    public void handle(String request) {
        if (request.contains("auth")) {
            System.out.println("AuthHandler: handled request");
        } else if (next != null) {
            next.handle(request);
        }
    }
}

class LoggerHandler extends Handler {

    @Override
    public void handle(String request) {
        System.out.println("Logger: " + request);
        if (next != null) {
            next.handle(request);
        }
    }
}

class BusinessHandler extends Handler {

    @Override
    public void handle(String request) {
        System.out.println("BusinessHandler: executing business logic");
    }
}

// Client
@SuppressWarnings("unused")
final class ChainOfResponsibilityExample {

    private ChainOfResponsibilityExample() {
    }

    public static void example() {
        Handler auth = new AuthHandler();
        Handler log = new LoggerHandler();
        Handler business = new BusinessHandler();
        auth.setNext(log).setNext(business);

        auth.handle("auth request");
        auth.handle("normal request");
    }
}

File: rust/src/behavioral/chain_of_responsibility.rs
/// middleware-style

trait Handler {
    fn set_next(&mut self, next: Option<Box<dyn Handler>>);
    fn handle(&self, req: &str);
}

struct BaseHandler {
    name: &'static str,
    next: Option<Box<dyn Handler>>,
}

impl BaseHandler {
    fn new(name: &'static str) -> Self {
        Self { name, next: None }
    }
}

impl Handler for BaseHandler {
    fn set_next(&mut self, next: Option<Box<dyn Handler>>) {
        self.next = next;
    }

    fn handle(&self, req: &str) {
        println!("{} received: {}", self.name, req);

        if let Some(ref next) = self.next {
            next.handle(req);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example() {
        let mut auth = BaseHandler::new("Auth");
        let mut log = BaseHandler::new("Logger");
        let biz = BaseHandler::new("Business");

        log.set_next(Some(Box::new(biz)));
        auth.set_next(Some(Box::new(log)));

        auth.handle("auth request");
        auth.handle("normal request");
    }
}
